rust/parquet/src/arrow/converter.rs

```diff
@@ -0,0 +1,293 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::arrow::record_reader::RecordReader;
+use crate::data_type::DataType;
+use arrow::array::Array;
+use std::convert::From;
+use std::slice::from_raw_parts;
+use std::sync::Arc;
+
+use crate::errors::Result;
+use arrow::array::BufferBuilder;
+use arrow::array::BufferBuilderTrait;
+use arrow::datatypes::ArrowPrimitiveType;
+
+use arrow::array::ArrayDataBuilder;
+use arrow::array::PrimitiveArray;
+use std::marker::PhantomData;
+use std::mem::transmute;
+
+use crate::data_type::BoolType;
+use crate::data_type::DoubleType as ParquetDoubleType;
+use crate::data_type::FloatType as ParquetFloatType;
+use crate::data_type::Int32Type as ParquetInt32Type;
+use crate::data_type::Int64Type as ParquetInt64Type;
+use arrow::datatypes::BooleanType;
+use arrow::datatypes::Float32Type;
+use arrow::datatypes::Float64Type;
+use arrow::datatypes::Int16Type;
+use arrow::datatypes::Int32Type;
+use arrow::datatypes::Int64Type;
+use arrow::datatypes::Int8Type;
+use arrow::datatypes::UInt16Type;
+use arrow::datatypes::UInt32Type;
+use arrow::datatypes::UInt64Type;
+use arrow::datatypes::UInt8Type;
+
+/// A converter is used to consume record reader's content and convert it to arrow
+/// primitive array.
+pub trait Converter<T: DataType> {
+    /// This method converts record reader's buffered content into arrow array.
+    fn convert(record_reader: &mut RecordReader<T>) -> Result<Arc<Array>>;
+}
+
+/// Trait for value to value conversion. This is similar to `std::convert::Into` in std
+/// lib. Due to the limitation of rust compiler, we can't implement trait
+/// `std::convert::Into` for primitives types.
+pub trait ConvertAs<T> {
+    /// This method does value to value conversion.
+    fn convert_as(self) -> T;
+}
+
+impl<T> ConvertAs<T> for T {
+    fn convert_as(self) -> T {
+        self
+    }
+}
+
+macro_rules! convert_as {
+    ($src_type: ty, $dest_type: ty) => {
+        impl ConvertAs<$dest_type> for $src_type {
+            fn convert_as(self) -> $dest_type {
+                self as $dest_type
+            }
+        }
+    };
+}
+
+convert_as!(i32, i8);
+convert_as!(i32, i16);
+convert_as!(i32, u8);
+convert_as!(i32, u16);
+convert_as!(i32, u32);
+
+/// Builder converter is used when parquet's data type is different from
+/// arrow's native type. In this case, we need to iterate over values in record reader
+/// and convert parquet value into arrow value one by one.
```

Comment:
Is the performance different if say you build an `Int32Array`, and then use the `cast` kernels to convert to an appropriate type? If a cast works out faster, then we could drop the `Converter` and remain with `DirectConverter`.

Reply:
Good point! Will try to use cast kernel.